lib/charms/sdcore_nms_k8s/v0/fiveg_core_gnb.py

```diff
@@ -0,0 +1,402 @@
+        super().__init__(handle)
+        self.tac = tac
+        self.plmns = plmns
+
+    def snapshot(self) -> dict:
+        """Return event data."""
+        return {
+            "tac": self.tac,
+            "plmns": self.plmns,
+        }
+
+    def restore(self, snapshot: dict) -> None:
+        """Restore event data.
+
+        Args:
+            snapshot (dict): contains information to be restored.
+        """
+        self.tac = snapshot["tac"]
+        self.plmns = snapshot["plmns"]
+
+
+class FivegCoreGnbRequirerAppData(BaseModel):
+    """Requirer application data for fiveg_core_gnb."""
+    cu_name: str = Field(
+        description="CU/gNB unique identifier",
+        examples=["gnb001"],
+    )
+
+
+class RequirerSchema(DataBagSchema):
+    """Requirer schema for fiveg_core_gnb."""
+
+    app_data: FivegCoreGnbRequirerAppData
+
+
+def data_matches_requirer_schema(data: dict) -> bool:
+    """Return whether data matches requirer schema.
+
+    Args:
+        data (dict): Data to be validated.
+
+    Returns:
+        bool: True if data matches requirer schema, False otherwise.
+    """
+    try:
+        RequirerSchema(app_data=FivegCoreGnbRequirerAppData(**data))
+        return True
+    except ValidationError as e:
+        logger.error("Invalid data: %s", e)
+        return False
+
+
+class FivegCoreGnbRequirerCharmEvents(CharmEvents):
+    """Custom events for the FivegCoreGnbRequirer."""
+
+    gnb_config_available = EventSource(GnbConfigAvailableEvent)
+
+
+class FivegCoreGnbRequires(Object):
+    """Class to be instantiated by requirer of the `fiveg_core_gnb`."""
+
+    on = FivegCoreGnbRequirerCharmEvents()  # type: ignore
+
+    def __init__(self, charm: CharmBase, relation_name: str, cu_name: str):
+        """Observes relation joined and relation changed events.
+
+        Args:
+            charm: Juju charm
+            relation_name (str): Relation name
+            cu_name (str): name of the CU/gNodeB
+        """
+        self.relation_name = relation_name
+        self.cu_name = cu_name
+        self.charm = charm
+        super().__init__(charm, relation_name)
+        self.framework.observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+
+    def publish_gnb_information(
+        self, relation_id: int, cu_name: str
+    ) -> None:
+        """Set CU/gNB identifier in the relation data.
+
+        Args:
+            relation_id (str): Relation ID.
+            cu_name (str): CU/gNB unique identifier.
+        """
+        if not data_matches_requirer_schema(
+            data={"cu_name": cu_name}
+        ):
+            raise ValueError(f"Invalid fiveG core gNB data: {cu_name}")
+        relation = self.model.get_relation(
+            relation_name=self.relation_name, relation_id=relation_id
+        )
+        if not relation:
+            raise RuntimeError(f"Relation {self.relation_name} not created yet.")
+        relation.data[self.charm.app]["cu_name"] = str(cu_name)
```

Comment:
why make the `cu_name` a string? didn't we validate already that it was a string?